File: typr-tester-typr-java/src/java/adventureworks/JsonEquals.java
package adventureworks;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

/**
 * JSON-based equality testing for Java tests, matching the Scala test approach. Uses Jackson
 * ObjectMapper to serialize objects to JSON and compare.
 */
public class JsonEquals {
  public static final ObjectMapper mapper =
      new ObjectMapper().registerModule(new Jdk8Module()).registerModule(new JavaTimeModule());

  public static void assertJsonEquals(Object expected, Object actual) {
    assertJsonEquals(null, expected, actual);
  }

  public static void assertJsonEquals(String message, Object expected, Object actual) {
    try {
      JsonNode expectedJson = mapper.valueToTree(expected);
      JsonNode actualJson = mapper.valueToTree(actual);
      if (message != null) {
        assertEquals(message, expectedJson, actualJson);
      } else {
        assertEquals(expectedJson, actualJson);
      }
    } catch (Exception e) {
      fail("Failed to serialize to JSON: " + e.getMessage());
    }
  }
}

File: typo-tester-mariadb-kotlin/src/kotlin/testdb/JsonEquals.kt
package testdb

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.datatype.jdk8.Jdk8Module
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.fasterxml.jackson.module.kotlin.KotlinModule
import org.junit.Assert.assertEquals
import org.junit.Assert.fail

object JsonEquals {
    val mapper: ObjectMapper = ObjectMapper()
        .registerModule(Jdk8Module())
        .registerModule(JavaTimeModule())
        .registerModule(KotlinModule.Builder().build())

    fun assertJsonEquals(expected: Any?, actual: Any?) {
        assertJsonEquals(null, expected, actual)
    }

    fun assertJsonEquals(message: String?, expected: Any?, actual: Any?) {
        try {
            val expectedJson: JsonNode = mapper.valueToTree(expected)
            val actualJson: JsonNode = mapper.valueToTree(actual)
            if (message != null) {
                assertEquals(message, expectedJson, actualJson)
            } else {
                assertEquals(expectedJson, actualJson)
            }
        } catch (e: Exception) {
            fail("Failed to serialize to JSON: ${e.message}")
        }
    }
}
